chromium_src/third_party/blink/renderer/modules/storage/dom_window_storage.cc

```diff
@@ -8,47 +8,160 @@
+  EphemeralStorageNamespaces(StorageController* controller,
+                             const String& session_storage_id,
+                             const String& local_storage_id);
+  virtual ~EphemeralStorageNamespaces() = default;
+
+  static const char kSupplementName[];
+  static EphemeralStorageNamespaces* From(Page* page);
+
+  StorageNamespace* session_storage() { return session_storage_.Get(); }
+  StorageNamespace* local_storage() { return local_storage_.Get(); }
+  void Trace(Visitor* visitor) const override;
+
+ private:
+  Member<StorageNamespace> session_storage_;
+  Member<StorageNamespace> local_storage_;
+};
+
+const char EphemeralStorageNamespaces::kSupplementName[] =
+    "EphemeralSessionStorageNamespaces";
+
+EphemeralStorageNamespaces::EphemeralStorageNamespaces(
+    StorageController* controller,
+    const String& session_storage_id,
+    const String& local_storage_id)
+    : session_storage_(
+          MakeGarbageCollected<StorageNamespace>(controller,
+                                                 session_storage_id)),
+      local_storage_(MakeGarbageCollected<StorageNamespace>(controller,
+                                                            local_storage_id)) {
+}
+
+void EphemeralStorageNamespaces::Trace(Visitor* visitor) const {
+  visitor->Trace(session_storage_);
+  visitor->Trace(local_storage_);
+  Supplement<Page>::Trace(visitor);
 }
 
-}  // namespace
+// static
+EphemeralStorageNamespaces* EphemeralStorageNamespaces::From(Page* page) {
+  if (!page)
+    return nullptr;
+
+  EphemeralStorageNamespaces* supplement =
+      Supplement<Page>::From<EphemeralStorageNamespaces>(page);
+  if (supplement)
+    return supplement;
+
+  // The ephemeral session storage namespace is constructed using the
+  // normal session storage id. We get it here and transform it.
+  StorageNamespace* session_storage_namespace = StorageNamespace::From(page);
+  if (!session_storage_namespace)
+    return nullptr;
 
-BraveDOMWindowStorage::BraveDOMWindowStorage() {}
-BraveDOMWindowStorage::~BraveDOMWindowStorage() {}
+  String session_storage_id = session_storage_namespace->namespace_id() +
+                              String("ephemeral-session-storage");
+  String local_storage_id = session_storage_namespace->namespace_id() +
+                            String("ephemeral-local-storage");
+  supplement = MakeGarbageCollected<EphemeralStorageNamespaces>(
+      StorageController::GetInstance(), session_storage_id, local_storage_id);
+
+  ProvideTo(*page, supplement);
+  return supplement;
+}
 
 // static
-StorageArea* BraveDOMWindowStorage::sessionStorage(LocalDOMWindow& window,
-                                              ExceptionState& exception_state) {
-  auto* storage =
-      DOMWindowStorage::From(window).sessionStorage(exception_state);
-  MaybeClearAccessDeniedException(storage, window, &exception_state);
-  return storage;
+const char BraveDOMWindowStorage::kSupplementName[] = "DOMWindowStorage";
```

Comment:
you can't use the same name as DOMWindowStorage for a different Supplement

Reply:
The documentation in `supplementable.h` seems to tell a different story:
```
// The Supplement is expected to use the same const char* string instance
// as its key. The Supplementable's SupplementMap will use the address of the
// string as the key and not the characters themselves. Hence, 2 strings with
// the same characters will be treated as 2 different keys.
```
Still, it's probably better to modify the string for clarity's sake. This should be fixed now.